### DISCUSSION TITLE
Virtual Env

### DISCUSSION THREAD

**Discussion by stefanbringuier** [marimo-team/marimo#221] [General]
How does one setup a Python virtual environment? Is it done prior to `pip install marimo` or does a `marimo` notebook have a mechanism for creating and manging virtual environments?

**Comment by akshayka** [comment]
Hi @stefanbringuier, welcome! Thanks for opening a discussion, 

Setting up a virtual environment is done prior to `pip install marimo`; `marimo` does not yet have a mechanism for managing virtual environments, but we may have one in the future (great idea!).

Instructions:

Create a virtual environment called "marimo-env" (you can name it whatever you like).
1. `python -m venv marimo-env`

Activate the environment (unix/macOS, let me know if you are running Windows):

2. `source marimo-env/bin/activate`

Install marimo

3. `pip install marimo`

When you're done working in the environment

4 `deactivate`

To learn more:
- Check out this excellent guide on scientific computing with Python: https://learn.scientific-python.org/development/tutorials/dev-environment/
- Read the official Python tutorial: https://docs.python.org/3/tutorial/venv.html 
- Or, ask me more questions :)

**Reply by stefanbringuier** [reply]
Thanks. I tried these steps but for example `import numpy as np` in a marimo notebook cell yields: 

```python
ModuleNotFoundError
This cell raised an exception: ModuleNotFoundError('No module named 'numpy'')

Tip:
See the console area for a traceback.
```

**Reply by akshayka** [reply]
That's because NumPy is not a required dependency of marimo. 

You can install numpy with `pip install numpy`. You might also want to get a couple plotting packages and a scientific computing package:
- `pip install matplotlib`
- `pip install altair`
- `pip install scipy`

marimo tries to have as few dependencies as possible, to minimize the chance of conflicting with the user's environment.

**Reply by akshayka** [reply]
Thanks for these comments by the way, they're very helpful. Keep them coming :) We'll make sure to update our docs to reflect these issues.

**Reply by stefanbringuier** [reply]
These are the steps I took, but I'm still getting issues:

```
python -m venv my_project
source my_project/bin/activate
pip install marimo
pip install numpy
pip install matplotlib

marimo edit my_notebook.py

Edit my_notebook.py in your browser 📝

        URL: http://localhost:2718
```
but I'm still getting an issue with importing modules like numpy.

**Reply by akshayka** [reply]
I wonder if you have multiple marimo installations, and it's picking up the wrong one.

Can you type `which marimo` and paste the output here?

EDIT: The output of `which marimo` should be `/path/to/my_project/bin/marimo`. If it's not, there's a path issue, in which case I would try

```
deactivate
pip uninstall marimo
source my_project/bin/activate
marimo edit
```

In particular, if you explicitly use `marimo` from your environment, that should (hopefully!) work:

```
source my_project/bin/activate
my_project/bin/marimo edit 
```

**Reply by stefanbringuier** [reply]
I think you are correct and I do have marimo installed at the system level.

```shell
$ which marimo
/usr/local/bin/marimo
```

It doesn't look like there is a  `my_project/bin/marimo` binary that I can add to my path via the `my_project/bin/activate` script. Any thoughts on a workaround?

**Reply by akshayka** [reply]
Hmm. I'd uninstall marimo from system level (if you haven't already), then try again from scratch.

```
$ deactivate
$ pip uninstall marimo
$ which marimo # sanity check, should be empty
$ rm -rf my_project/ # destroy the environment
$ python -m venv my_project # create it again from scratch
$ source my_project/bin/activate
$ pip install marimo
$ which marimo # sanity check, shouldn't be empty
$ pip install numpy
$ pip install matplotlib
$ marimo edit
```

Sorry, Python environment management can be a bit cumbersome. A good argument for marimo to abstract it away ...!

**Reply by stefanbringuier** [reply]
Did the trick. 

I was hoping to have a "vanilla" marimo setup at the system level for quick stuff, but it's probably safer and cleaner otherwise.

Yeah if marimo can abstract and handle it, that would be awesome. I would probably never touch Jupyter notebooks again.

**Reply by akshayka** [reply]
Awesome, glad to hear it!

> Yeah if marimo can abstract and handle it, that would be awesome. I would probably never touch Jupyter notebooks again.

Out of curiosity, how do you deal with virtual environments for Jupyter today? Through VS Code? Or do you just install your packages system wide

**Comment by s-celles** [comment]
Hi,

Daytona seems an interesting approach to consider for building development environment.

Daytona.io is an open-source Development Environment Manager designed to simplify and streamline the process of setting up development environments[1][3]. Key features of Daytona include:

1. Single-command setup: Developers can create a full dev environment using just one command: 'daytona create'[1].

2. Flexibility: It works on any machine and with any infrastructure, supporting both local and remote environments[1][3].

3. Automation: Daytona automates the creation of development environments, connects to local instances of VS Code (or browser-based alternatives), creates VPN tunnels, and adds reverse proxies for sharing[1].

4. Standardization: It helps eliminate configuration drift and ensures consistent environments throughout the development lifecycle[3].

5. Scalability: Daytona can scale with development requirements, removing constraints of local machines[3].

6. Security: It offers fine-grained code access controls and comprehensive dependency management[3].

7. Integration: Daytona supports various Git providers (GitHub, GitLab, Bitbucket) and popular IDEs (Visual Studio Code, JetBrains, Vim)[3].

Daytona.io has gained significant traction since its launch, amassing over 6,700 GitHub stars[2]. The company has also secured $7 million in funding to advance its mission of simplifying development environments[2][4].

Daytona (can) rely on Docker as provider.

https://www.youtube.com/watch?v=uL-TaEhvVwk provides an introduction to this tool

https://www.youtube.com/watch?v=DmLj1fTQebw&t=30s provides an example with Jupyter Notebooks in Seconds - No Setup, Remote Access, GPU-Ready... It could probably be Marimo.

I wonder if some of you have ever experiment it and have an (unbiased) opinion about it.

Citations:
[1] https://news.ycombinator.com/item?id=39616709
[2] https://www.youtube.com/watch?v=RNf_vfgc91c
[3] https://www.daytona.io
[4] https://www.linkedin.com/company/daytonaio
[5] https://github.com/daytonaio/daytona/actions
[6] https://twitter.com/daytonaio
[7] https://dribbble.com/SLRNCL/projects/7008631-Daytona-io
[8] https://x.com/daytonaio/highlights

Best regards
